game-state-service

```diff
@@ -0,0 +1,157 @@
+
+
+class GameStateService(object):
+
+    _DBUS_NAME = 'com.endlessm.GameStateService'
+    _DBUS_IFACE = 'com.endlessm.GameStateService'
+    _DBUS_PATH = '/com/endlessm/GameStateService'
+    _DBUS_KEY_ERROR = 'com.endlessm.GameStateService.KeyError'
+    _DBUS_KEY_MESSAGE = 'key %s does not exist'
+    _DBUS_SIGNATURE_ERROR = 'com.endlessm.GameStateService.SignatureError'
+    _DBUS_SIGNATURE_MESSAGE = 'signature %s does not match key %s'
+    _DBUS_XML = '''
+    <node>
+      <interface name='com.endlessm.GameStateService'>
+        <method name='Get'>
+          <arg type='s' name='key' direction='in'/>
+          <arg type='s' name='signature' direction='in'/>
+          <arg type='v' name='value' direction='out'/>
+        </method>
+        <method name='Set'>
+          <arg type='s' name='key' direction='in'/>
+          <arg type='v' name='value' direction='in'/>
+        </method>
+        <signal name='changed'>
+          <arg type='s' name='key'/>
+          <arg type='v' name='value'/>
+        </signal>
+      </interface>
+    </node>
+    '''
+
+    def __init__(self):
+        self._loop = GLib.MainLoop()
+        self._node = Gio.DBusNodeInfo.new_for_xml(self._DBUS_XML)
+        self._info = self._node.lookup_interface(self._DBUS_IFACE)
+        self._owner_id = Gio.bus_own_name(Gio.BusType.SESSION,
+                                          self._DBUS_NAME,
+                                          Gio.BusNameOwnerFlags.NONE,
+                                          self._on_bus_acquired,
+                                          self._on_name_acquired,
+                                          self._on_name_lost)
+
+    def _on_method_called(self, connection, sender, path, interface,
+                          method, parameters, invocation):
+        if method == 'Set':
+            key, value = parameters.unpack()
+            self._state.set(key, value)
+            assert connection.emit_signal(None,
+                                          self._DBUS_PATH,
+                                          self._DBUS_IFACE,
+                                          'changed',
+                                          parameters)
+            invocation.return_value(None)
+        elif method == 'Get':
+            key, signature = parameters.unpack()
+            try:
+                value = self._state.get(key)
+                variant = GLib.Variant(signature, value)
+            except KeyError:
+                invocation.return_dbus_error(
+                    self._DBUS_KEY_ERROR,
+                    self._DBUS_KEY_MESSAGE % key)
+            except (TypeError, IndexError):
+                invocation.return_dbus_error(
+                    self._DBUS_SIGNATURE_ERROR,
+                    self._DBUS_SIGNATURE_MESSAGE % (signature, key))
+            else:
+                invocation.return_value(GLib.Variant('(v)', (variant,)))
+
+    def _on_bus_acquired(self, connection, name):
```

Comment:
You can probably make this less verbose by inheriting from `Gio.Application` and implementing the `do_dbus_register` and `do_dbus_unregister` virtual methods.